### PR TITLE
feat: add new command restore-database-mapping

### LINKED ISSUE #269: FR: Ability to restore to another database.
We need to restore the same backup 20 times into the same server but into different databases.

```
clickhouse-backup restore bkp01 -c config.yml --remap-schema databaseX:databaseX1,databaseY:databaseY1
clickhouse-backup restore bkp01 -c config.yml --remap-schema databaseX:databaseX2,databaseY:databaseY2
```

original databaseX will restored into databaseX1 and databaseX2

### PR DESCRIPTION
**Pull request by mojerro**
fix #269 

add new optional command `restore-database-mapping` to support restoring tables into specific databases.

### CODE REVIEW

**Comment by mojerro** [comment]
3 todos and unit tests left, continuously learning the code structure now

```
TODO(mojerro): if restore-database-mapping is specified, create databases in mapping rules.
```

File: ReadMe.md [hunk 1]
```diff
@@ -109,6 +109,7 @@ general:
   restore_schema_on_cluster: ""  # RESTORE_SCHEMA_ON_CLUSTER, execute all schema related SQL queryes with `ON CLUSTER` clause as Distributed DDL, look to `system.clusters` table for proper cluster name
   upload_by_part: true           # UPLOAD_BY_PART
   download_by_part: true         # DOWNLOAD_BY_PART
+  restore_database_mapping: ""   # RESTORE_DATABASE_MAPPING, restore rules from backup databases to target databases, which is useful on change the restore database or restore part of the backup databases.
```
Comment (Slach) [inline ReadMe.md]: look to disk_mapping
`{}`

properly empty value

File: cmd/clickhouse-backup/main.go [hunk 2]
```diff
@@ -231,16 +231,21 @@ func main() {
 		{
 			Name:      "restore",
 			Usage:     "Create schema and restore data from backup",
-			UsageText: "clickhouse-backup restore  [-t, --tables=<db>.<table>] [--partitions=<partitions_names>] [-s, --schema] [-d, --data] [--rm, --drop] [--rbac] [--configs] <backup_name>",
+			UsageText: "clickhouse-backup restore  [-t, --tables=<db>.<table>] [-m, --map=<originDB>:<targetDB>[,...]] [--partitions=<partitions_names>] [-s, --schema] [-d, --data] [--rm, --drop] [--rbac] [--configs] <backup_name>",
```
Comment (Slach) [inline cmd/clickhouse-backup/main.go]: wrong non mental common name
use
--restore-disk-mapping as the same as in config
you can add aliases

File: cmd/clickhouse-backup/main.go [hunk 3]
```diff
@@ -231,16 +231,21 @@ func main() {
...
+				return backup.Restore(config.GetConfig(c), c.Args().First(), c.String("t"), c.StringSlice("map"), c.StringSlice("partitions"), c.Bool("s"), c.Bool("d"), c.Bool("rm"), c.Bool("rbac"), c.Bool("configs"))
 			},
 			Flags: append(cliapp.Flags,
 				cli.StringFlag{
 					Name:   "table, tables, t",
 					Usage:  "table name patterns, separated by comma, allow ? and * as wildcard",
 					Hidden: false,
 				},
+				cli.StringFlag{
+					Name: "map, restore-database-mapping, m",
```
Comment (Slach) [inline cmd/clickhouse-backup/main.go]: replace StringFlag to StringSlice,if you want use it as string slice

File: cmd/clickhouse-backup/main.go [hunk 4]
```diff
@@ -276,17 +281,22 @@ func main() {
...
-				return b.RestoreFromRemote(c.Args().First(), c.String("t"), c.StringSlice("partitions"), c.Bool("s"), c.Bool("d"), c.Bool("rm"), c.Bool("rbac"), c.Bool("configs"))
+				return b.RestoreFromRemote(c.Args().First(), c.String("t"), c.StringSlice("map"), c.StringSlice("partitions"), c.Bool("s"), c.Bool("d"), c.Bool("rm"), c.Bool("rbac"), c.Bool("configs"))
 			},
 			Flags: append(cliapp.Flags,
 				cli.StringFlag{
 					Name:   "table, tables, t",
 					Usage:  "table name patterns, separated by comma, allow ? and * as wildcard",
 					Hidden: false,
 				},
+				cli.StringFlag{
```
Comment (Slach) [inline cmd/clickhouse-backup/main.go]: replace StringFlag to StringSlice,if you want use it as string slice

File: pkg/backup/restore.go [hunk 5]
```diff
@@ -253,6 +262,7 @@ func createTables(cfg *config.Config, ch *clickhouse.ClickHouse, tablesForRestor
 		var notRestoredTables ListOfTables
 		for _, schema := range tablesForRestore {
 			// if metadata.json doesn't contains "databases", we will re-create tables with default engine
+			// TODO(mojerro): if restore-database-mapping specified, create database in mapping rules.
```
Comment (Slach) [inline pkg/backup/restore.go]: please resolve this TODO
we shall create databases properly

File: pkg/backup/restore.go [hunk 6]
```diff
@@ -263,6 +273,7 @@ func createTables(cfg *config.Config, ch *clickhouse.ClickHouse, tablesForRestor
 			schema.Query = strings.Replace(
 				schema.Query, "CREATE WINDOW VIEW", "ATTACH WINDOW VIEW", 1,
 			)
+			// TODO(mojerro): if restore-database-mapping specified, use mapping database
```
Comment (Slach) [inline pkg/backup/restore.go]: just implements ch.getDatabaseNameWithMapping(schema.Database)

File: pkg/server/server.go [hunk 7]
```diff
@@ -784,6 +785,10 @@ func (api *APIServer) httpRestoreHandler(w http.ResponseWriter, r *http.Request)
 		tablePattern = tp[0]
 		fullCommand = fmt.Sprintf("%s --tables=\"%s\"", fullCommand, tablePattern)
 	}
+	if databaseMapping, exist := query["databaseMapping"]; exist {
```
Comment (Slach) [inline pkg/server/server.go]: please use the same name as config
`restore_database_mapping` looks more mental we don't use "camelCase" in other query string parameters as you see

File: pkg/server/server.go [hunk 8]
```diff
@@ -784,6 +785,10 @@ func (api *APIServer) httpRestoreHandler(w http.ResponseWriter, r *http.Request)
...
+		databaseMappingToBackup = strings.Split(databaseMapping[0], ",")
```
Comment (Slach) [inline pkg/server/server.go]: databaseMappingToRestore
more mental name here

**Review by Slach — CHANGES_REQUESTED** [review]
i don't see real database name replace, just add paremeter
please resolve comments and implements something like
`clickhouse.GetDatabaseFromMapingToRestore`

and use it inside restoreSchema and restoreData methods

**Comment by Slach** [comment]
moreover
tesflows test failed
https://github.com/AlexAkulov/clickhouse-backup/runs/6144342294?check_suite_focus=true
something wrong with config marshalling

```
                 3ms                [clickhouse_backup] bash# clickhouse-backup delete local bp_all_dt
                78ms                [clickhouse_backup] 2022/04/24 07:04:02.242509 error can't parse config file: yaml: unmarshal errors:
                78ms                [clickhouse_backup]   line 75: cannot unmarshal !!str `default...` into map[string]string
                82ms                [clickhouse_backup] bash# echo $?
                82ms                [clickhouse_backup] 1
```
install docker-compose and python3 
and run tests
look to https://github.com/AlexAkulov/clickhouse-backup/blob/master/test/testflows/README.md for details

**Comment by mojerro** [comment]
No, I'm not implementing all code now, just creating a draft. I will change to review after I finish. Thanks for your review and advice~

**Comment by mojerro** [comment]
one test error left

```
Apr 27,2022 18:45:00           ⟥  By executing command
                                    echo -e "SELECT * FROM mysql.MyTable" | clickhouse client -n
                 5ms                [clickhouse1] bash# echo -e "SELECT * FROM mysql.MyTable" | clickhouse client -n
                45ms                [clickhouse1] Received exception from server (version 22.3.4):
                45ms                [clickhouse1] Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table mysql.MyTable doesn't exist. (UNKNOWN_TABLE)
                45ms                [clickhouse1] (query: SELECT * FROM mysql.MyTable
                45ms                [clickhouse1] )
                54ms                [clickhouse1] bash# echo $?
                54ms                [clickhouse1] 60
                56ms                [clickhouse1] bash#
                56ms           ⟥⟤ OK executing command, /clickhouse backup/other engines/materializedmysql/I create MaterializedMySQL/executing command
Apr 27,2022 18:45:00           ⟥  Then check if output has exception
                19ms           ⟥    Exception: AssertionError: Oops! Assertion failed

                                      Description
                                        Received exception from server (version 22.3.4):
                                      Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table mysql.MyTable doesn't exist. (UNKNOWN_TABLE)
                                      (query: SELECT * FROM mysql.MyTable
                                      )
                19ms           ⟥⟤ Fail check if output has exception, /clickhouse backup/other engines/materializedmysql/I create MaterializedMySQL/check if output has exception, AssertionError
                                 AssertionError: Oops! Assertion failed

                                 Description
                                   Received exception from server (version 22.3.4):
                                 Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table mysql.MyTable doesn't exist. (UNKNOWN_TABLE)
                                 (query: SELECT * FROM mysql.MyTable
                                 )
               149ms         ⟥⟤ XFail I create MaterializedMySQL, /clickhouse backup/other engines/materializedmysql/I create MaterializedMySQL, AssertionError
                               AssertionError: Oops! Assertion failed

                               Description
                                 Received exception from server (version 22.3.4):
                               Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table mysql.MyTable doesn't exist. (UNKNOWN_TABLE)
                               (query: SELECT * FROM mysql.MyTable
                               ), DROP TABLE not supported by MaterializedMySQL, just attach will not help
```

image

**Comment by Slach** [comment]
hmm, `DROP TABLE not supported by MaterializedMySQL, just attach will not help` should be XFail (expected fail) and testflows test shall pass

File: pkg/backup/restore.go [hunk 9]
```diff
@@ -231,6 +240,21 @@ func RestoreSchema(cfg *config.Config, ch *clickhouse.ClickHouse, backupName str
 	if err != nil {
 		return err
 	}
+	// if restore-database-mapping specified, create database in mapping rules instead of in backup files.
+	if len(dbMapRule) > 0 {
+		err = getTableListByRestoreDatabaseMappingRule(&tablesForRestore, dbMapRule)
+		// println("************************************")
```
Comment (Slach) [inline pkg/backup/restore.go]: please remove unused code

File: pkg/backup/table_pattern.go [hunk 10]
```diff
@@ -109,6 +109,31 @@ func getTableListByPatternLocal(metadataPath string, tablePattern string, skipTa
...
+	dbMetaMap := make(map[string]metadata.TableMetadata, 0)
+	for i := 0; i < len(*originTables); i++ {
+		table := (*originTables)[i]
+		dbMetaMap[table.Database] = table
+	}
+	for i := 0; i < len(*originTables); i++ {
+		originDBMeta := (*originTables)[i]
+		if targetDB, ok := dbMapRule[originDBMeta.Database]; ok {
+			originDBMeta.Query = strings.Replace(
+				originDBMeta.Query, originDBMeta.Database+".", targetDB+".", 1,
```
Comment (Slach) [inline pkg/backup/table_pattern.go]: we have some corner cases
when 
database name escaped with **`** character
also we shall avoid corner cases when .Query contains something like
```CREATE VIEW db1.view AS SELECT * FROM db2_db1.another_table```

i propose replace database name with regexp something like that
```golang
regExp := `^CREATE (TABLE|VIEW|MATERIALIZED VIEW|DICTiONARY|FUNCTION) ([`]*)(db)([`]*)\.`
toStr := "CREATE $1 $2new_db$4."
```

File: test/testflows/clickhouse_backup/configs/backup/config.yml [hunk 11]
```diff
@@ -72,6 +72,9 @@ general:
   max_file_size: 107374182400
   remote_storage: none
   upload_concurrency: 1
+  restore_database_mapping: 
+    default: default
```
Comment (Slach) [inline test/testflows/clickhouse_backup/configs/backup/config.yml]: we should make another test case instead modify base config
and check 

if you want just cover this case inside ./test/integration/integration_test.go`
i could be easier modify for you

**Comment by Slach** [comment]
```
[ Fail ] /clickhouse backup/cli/default config
```
try this
```bash
rm -rf./test/testflows/clickhouse_backup/tests/snapshots/cli.py.cli.snapshot
python3 ./test/testflows/clickhouse_backup/regression.py --only "/clickhouse backup/cli/default config*"
```
should pass and create new *.snapshot file

**Review by Slach — CHANGES_REQUESTED** [review]

**Review by Slach — CHANGES_REQUESTED** [review]
look to https://github.com/AlexAkulov/clickhouse-backup/pull/432/files#diff-59718d24adef263b3b224c44b21735795ec8e96b952e96724a81e59500dff19bR73

we need go add apply restore database mapping  during create empty databases

File: test/testflows/clickhouse_backup/configs/backup/config.yml (hunk 11 above)
Comment (mojerro) [reply test/testflows/clickhouse_backup/configs/backup/config.yml]: I can try it out first

**Comment by mojerro** [comment]
> look to https://github.com/AlexAkulov/clickhouse-backup/pull/432/files#diff-59718d24adef263b3b224c44b21735795ec8e96b952e96724a81e59500dff19bR73
> 
> we need go add apply restore database mapping during create empty databases

I'm not sure what you mean.  Running `clickhouse-backup restore --schema --drop <backup_name>` will create a empty database, do you mean that? Or which function needs to deal with.

**Comment by Slach** [comment]
> I'm not sure what you mean. Running `clickhouse-backup restore --schema --drop <backup_name>` will create a empty database, do you mean that? Or which function needs to deal with.

i mean
```yaml
 restore_database_mapping: 
    default: default
```

have no sense and do nothing
we need separate test case where we will restore with apply restore_database_mapping

**Comment by mojerro** [comment]
oh, yes, there are wroks to be done with test cases.

**Comment by mojerro** [comment]
I have a question, shall we deal with the `ATTACH` command in `metadata.json`.

`metrics_counters_local.json`
```json
{
	"table": "metrics_counters_local",
	"database": "default",
	"parts": {},
	"query": "CREATE TABLE default.metrics_counters_local (`org_id` UInt64, `project_id` UInt64, `metric_id` UInt64, `granularity` UInt32, `tags.key` Array(UInt64), `tags.value` Array(UInt64), `_tags_hash` Array(UInt64) MATERIALIZED arrayMap((k, v) -\u003e cityHash64(concat(toString(k), '=', toString(v))), tags.key, tags.value), `timestamp` DateTime, `retention_days` UInt16, `value` AggregateFunction(sum, Float64),  INDEX bf_tags_hash _tags_hash TYPE bloom_filter() GRANULARITY 1,  INDEX bf_tags_key_hash tags.key TYPE bloom_filter() GRANULARITY 1) ENGINE = AggregatingMergeTree() PARTITION BY (retention_days, toMonday(timestamp)) ORDER BY (org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value) SETTINGS index_granularity = 256",
	"size": {},
	"metadata_only": false
}
```
`metrics_counters_mv_10s_local.json`
```json
{
	"table": "metrics_counters_mv_10s_local",
	"database": "default",
	"parts": null,
	"query": "ATTACH MATERIALIZED VIEW default.metrics_counters_mv_10s_local TO default.metrics_counters_local (`org_id` UInt64, `project_id` UInt64, `metric_id` UInt64, `granularity` UInt32, `tags.key` Array(UInt64), `tags.value` Array(UInt64), `timestamp` DateTime, `retention_days` UInt16, `value` AggregateFunction(sum, Float64)) AS SELECT org_id, project_id, metric_id, 10 AS granularity, tags.key, tags.value, toStartOfInterval(timestamp, toIntervalSecond(10)) AS timestamp, retention_days, sumState(value) AS value FROM default.metrics_counters_buckets_local WHERE materialization_version = 0 GROUP BY org_id, project_id, metric_id, tags.key, tags.value, timestamp, granularity, retention_days",
	"size": null,
	"metadata_only": false
}
```

for now, I didn't deal with attach command
```go
regExp := regexp.MustCompile(`(?m)^CREATE (TABLE|VIEW|MATERIALIZED VIEW|DICTiONARY|FUNCTION) ([\x60]?)([^\x60]*)([\x60]?)\.`)
substitution := fmt.Sprintf("CREATE ${1} ${2}%v${4}.", targetDB)

originDBMeta.Query = regExp.ReplaceAllString(originDBMeta.Query, substitution)
originDBMeta.Database = targetDB
```

**Comment by mojerro** [comment]
I think I should deal with it : )

**Comment by Slach** [comment]
yes, you need support ATTACH MATERIALIZED VIEW

**Comment by mojerro** [comment]
@Slach Should I deal with the function `GetBackupTablesLegacy`?

```go
var tablesForRestore ListOfTables
	if backup.Legacy {
		tablesForRestore, err = ch.GetBackupTablesLegacy(backupName, disks)
```


I'm not sure if I can ignore it

**Comment by Slach** [comment]
I think we can ignore legacy 0.x backup version for this function
or just raise fatal error here too much refactoring
we plan remove "legacy" backup support in 2.0.x versions

**Comment by mojerro** [comment]
that's fine

**Comment by mojerro** [comment]
Also, how can I debug with vscode, if I run `go run ./...` directly, it will show some unexpected error

**Comment by Slach** [comment]
try to `go run ./cmd/clickhouse-backup/main.go`

sorry, i don't use vscode

**Comment by mojerro** [comment]
Well, that's ok. When I run `go run ./cmd/clickhouse-backup/main.go` directly, it will shows error
```
2022/05/12 14:47:04.997629 error can't attach partitions for table 'default.errors_local': code: 233, message: Detached part "90-20220228_1_59_17" not found
exit status 1
```

but everything goes well after I compile  it to docker.

**Comment by mojerro** [comment]
I'm using WSL2, the problem may come from this.

**Comment by Slach** [comment]
do you run vscode inside WSL2 ?
vscode run windows go .exe run ?

**Comment by mojerro** [comment]
yes, I'm running vscode in WSL2 with RemoteWSL extension,

**Comment by Slach** [comment]
ok. "remoteWSL" means not "inside WSL" you run windows binary vscode.exe and it run `go run ...` inside WSL

ok. do you run clickhouse-server as systemctl service inside WSL? 
or you run clickhouse-server inside docker which run inside the same WSL where you run `go run`? 
or you run clickhouse-server inside separate WSL instance which managed by your windows Docker-Desktop?

**Comment by mojerro** [comment]
I have a check, and I run clickhouse-server inside docker which run inside the same WSL where I run `go run`
Or I should start a local clickhouse-server in wsl for develop?

By the way, how can I have notification when there are new messages

**Comment by Slach** [comment]
@mojerro 
`clickhouse-backup` require direct access to `/var/lib/clickhouse` folder which used by `clickhouse-server`
if you run a docker container inside WSL, then you need to mount volume inside `docker container` to `/var/lib/clickhouse`

something like this
in separate WSL bash window
```bash
mkdir -p /var/lib/clickhouse 
docker run --name clickhouse -p 9000:9000 -v /var/lib/clickhouse/:/var/lib/clickhouse clickhouse/clickhouse-server:latest
```

after it `go run ./cmd/clickhouse-backup/main.go restore ... ` shall works as expected

**Comment by Slach** [comment]
> By the way, how can I have notification when there are new messages
@mojerro 
you should be subscribed to any changes and messages to this PR
look to attached screenshot
image

**Comment by mojerro** [comment]
> @mojerro `clickhouse-backup` require direct access to `/var/lib/clickhouse` folder which used by `clickhouse-server` if you run a docker container inside WSL, then you need to mount volume inside `docker container` to `/var/lib/clickhouse`
> 
> something like this in separate WSL bash window
> 
> ```shell
> mkdir -p /var/lib/clickhouse 
> docker run --name clickhouse -p 9000:9000 -v /var/lib/clickhouse/:/var/lib/clickhouse clickhouse/clickhouse-server:latest
> ```
> 
> after it `go run ./cmd/clickhouse-backup/main.go restore ... ` shall works as expected

Thanks for your explanation : )

**Comment by Slach** [comment]
@mojerro do you have any progress? i planning release 1.4.0 today, could you complete your PR ?

**Comment by mojerro** [comment]
The code can run successfully now, but I haven't added a test case and my code's structure is not brief enough.

File: pkg/backup/restore.go [hunk 12]
```diff
@@ -69,9 +71,21 @@ func Restore(cfg *config.Config, backupName, tablePattern string, databaseMappin
 		}
 		if schemaOnly || doRestoreData {
 			for _, database := range backupMetadata.Databases {
-				if !IsInformationSchema(database.Name) {
-					if err := ch.CreateDatabaseFromQuery(database.Query); err != nil {
-						return err
+				if targetDB, ok := dbMapRule[database.Name]; ok && len(dbMapRule) > 0 {
```
Comment (Slach) [inline pkg/backup/restore.go]: len(dbMapRule) always > 0 here

Comment (mojerro) [reply pkg/backup/restore.go]: 😅

File: pkg/backup/restore.go [hunk 13]
```diff
@@ -69,9 +71,21 @@ func Restore(cfg *config.Config, backupName, tablePattern string, databaseMappin
...
+					// When create database, try to substitute the database by following the database mapping rule.
+					if !IsInformationSchema(targetDB) {
+						regExp := regexp.MustCompile(`(?m)^CREATE DATABASE ([\x60]?)([^\x60]*)([\x60]?)`)
```
Comment (Slach) [inline pkg/backup/restore.go]: I think we could move this compilation into `pkg/clickhouse/clickhouse.go`, something like `var createDatabaseRE = regexp.MustCompile()` 

and use it as `clickhouse.createDatabaseRE.ReplaceAllString(...)` in code

File: pkg/backup/restore.go [hunk 14]
```diff
@@ -4,28 +4,39 @@ import (
...
+
 	"github.com/mattn/go-shellwords"
 
 	"github.com/AlexAkulov/clickhouse-backup/pkg/clickhouse"
 	"github.com/AlexAkulov/clickhouse-backup/pkg/filesystemhelper"
 	"github.com/AlexAkulov/clickhouse-backup/pkg/metadata"
 	"github.com/AlexAkulov/clickhouse-backup/pkg/utils"
 	apexLog "github.com/apex/log"
-	"github.com/otiai10/copy"
+	otiai10_copy "github.com/otiai10/copy"
```
Comment (Slach) [inline pkg/backup/restore.go]: `recursive_copy` here will more mental instead of `otai10_copy`

**Review by Slach — CHANGES_REQUESTED** [review]
looks good in golang part
please add test case https://github.com/AlexAkulov/clickhouse-backup/pull/432/files#r859731218

and fix other review comments

**Comment by mojerro** [comment]
I'm learning about how to use testflow and may use it some days. This is my first Pr, still learning about coding ✈️

**Comment by mojerro** [comment]
Sorry about my low-speed coding.

**Comment by Slach** [comment]
All ok, I appreciate for your efforts! Your coding speed is not low

File: test/testflows/clickhouse_backup/configs/backup/config.yml (hunk 11 above)
Comment (Slach) [reply test/testflows/clickhouse_backup/configs/backup/config.yml]: your changes works, but broke testflows
look 
https://github.com/AlexAkulov/clickhouse-backup/runs/6466446682?check_suite_focus=true
for details

**Comment by Slach** [comment]
@mojerro do you have any progress?

**Comment by mojerro** [comment]
I meet some problems with my test environment, I can start the docker compose ,but it shows a timeout error. Still trying to fix it.

**Comment by mojerro** [comment]
Everything went well a couple of weeks ago. It shows the same error even if I switched to the master branch.

**Comment by Slach** [comment]
try to 
`docker-compose down`
before start

**Comment by mojerro** [comment]
image
I met lots of different problem, and I solve some of them by using break point. But still some problems made me puzzle. I will try to develop on my linux server.

**Comment by mojerro** [comment]
this is not the final code, you can review it later.

**Comment by Slach** [comment]
@mojerro 
```bash
Error: pkg/backup/restore.go:11:2: imported and not used: "regexp"
```

let's fix it?

**Comment by Slach** [comment]
@mojerro do you will continue work with this PR?

**Comment by Slach** [comment]
@mojerro don't be shy, your changes almost done, you just need some small fixes in test cases and in codebase

**Comment by mojerro** [comment]
ok, I'll have a try this weekend.  I don't know why the test case was suddenly false and spent lots of time on it. That makes me feel frustrated.

File: pkg/backup/table_pattern.go [hunk 15]
```diff
@@ -109,6 +111,58 @@ func getTableListByPatternLocal(metadataPath string, tablePattern string, skipTa
...
+				substitution = fmt.Sprintf("${1} ${2} ${3}%v${5}.${6}${7}${8}%v${11}", targetDB, targetDB)
+			} else {
+				return fmt.Errorf("there is an error when substituting the default database in query, try to open an issue on Github and we will fix it :)")
+			}
+
+			originDBMeta.Query = regExp.ReplaceAllString(originDBMeta.Query, substitution)
+			originDBMeta.Database = targetDB
+			result = append(result, originDBMeta)
+		}
+		// FIXME: Need to optimize
```
Comment (Slach) [inline pkg/backup/table_pattern.go]: Could you clarify, what exactly this part of code do?

`table.Parts` doesn't split by `database` it splitted by diskname

**Review by Slach — CHANGES_REQUESTED** [review]

File: pkg/backup/restore.go [hunk 16]
```diff
@@ -367,7 +405,30 @@ func RestoreData(cfg *config.Config, ch *clickhouse.ClickHouse, backupName strin
 	if err != nil {
 		return err
 	}
+	var originDisks = make([]clickhouse.Disk, len(disks))
+	copy(originDisks, disks)
 	diskMap := map[string]string{}
+	if len(dbMapRule) > 0 {
+		var _disks []clickhouse.Disk
+		for i := range disks {
+			if name, ok := dbMapRule[disks[i].Name]; ok {
```
Comment (Slach) [inline pkg/backup/restore.go]: Why do you try to use `dbMapRule` and change disks from system.disks?

I can't  understand sense of this part of code
we already have in config `clickhouse->disk_mapping`

Use `--restore-database-mapping` as `--restore-disk-mapping` broke single responsibility princile for this parameter

**Review by Slach — CHANGES_REQUESTED** [review]

File: pkg/backup/table_pattern.go [hunk 17]
```diff
@@ -109,6 +111,58 @@ func getTableListByPatternLocal(metadataPath string, tablePattern string, skipTa
 	return result, nil
 }
 
+func getTableListByRestoreDatabaseMappingRule(originTables *ListOfTables, dbMapRule map[string]string) error {
+	result := ListOfTables{}
+	dbMetaMap := make(map[string]metadata.TableMetadata, 0)
+	for i := 0; i < len(*originTables); i++ {
+		table := (*originTables)[i]
+		dbMetaMap[table.Database] = table
```
Comment (Slach) [inline pkg/backup/table_pattern.go]: dbMetaMap doesn't use in getTableListByRestoreDatabaseMappingRule

**Review by Slach — APPROVED** [review]

File: pkg/clickhouse/clickhouse.go [hunk 18]
```diff
@@ -713,22 +713,28 @@
 }
 
 func (ch *ClickHouse) Query(query string, args ...interface{}) (sql.Result, error) {
-	return ch.conn.Exec(ch.LogQuery(query), args...)
+	return ch.conn.Exec(ch.LogQuery(query, args...), args...)
 }
 
 func (ch *ClickHouse) Queryx(query string, args ...interface{}) (*sqlx.Rows, error) {
-	return ch.conn.Queryx(ch.LogQuery(query), args...)
+	return ch.conn.Queryx(ch.LogQuery(query, args...), args...)
```
[the remark this answers is not in the document]
Comment (Slach) [reply pkg/clickhouse/clickhouse.go:720]: false positive

**Comment by Slach** [comment]
@mojerro Thanks for contribution
I hope your next PR will great ;)

**Pull request merged**